File: php/Signature.php
<?php

class Signature{

  private $secret;

  function __construct($sharedSecret){
    $this->secret = $sharedSecret;
  }

  private function validateRequest($request){
    if(!(is_array($request) && array_key_exists("signature",$request))) return false;
    $signature = $request['signature'];
    if(!(array_key_exists('signed',$signature)&&array_key_exists('token',$signature)&&array_key_exists('nonce',$signature)))return false;
    return true;
  }

  private function recreateToken($request, $keys){
      $signingString = "";
      foreach ($keys as $key) {
        $signingString .= @$request[$key];
      }
      $signingString .= $request['signature']['nonce'] . $this->secret;
      return sha1($signingString);
  }

  public function validateSignature($request){
      if (!($this->validateRequest($request)))return false;
      $keys = preg_split("/,/",$request['signature']['signed']);
      return $request['signature']['token'] == $this->recreateToken($request,$keys);
  }

  public function generateSignature(&$requestArray,$nonce){
      $signed = $signingString ="";
      foreach ($requestArray as $k => $v) {
        $signed .= $k;
        $signingString .= $v;
      }
      $token = sha1($signingString . $nonce . $this->secret);
      $requestArray['signature'] = array(
                                        "nonce" => $nonce,
                                        "signed" => $signed,
                                        "token" => $token
                                  );
  }
}

File: python/Signature.py
from hashlib import sha1

class Signature() :

    def __init__(self,secret):
        self.secret = secret

    def validate(self,request):
        keys = request['signature']['signed'].split(',')
        signing_string = ''
        for key in keys :
            signing_string += request[key]
        token = sha1(signing_string + request['signature']['nonce'] + self.secret).hexdigest()
        return token == request['signature']['token']

    def generate(self,request,nonce):
        signed = []
        signing_string = ""
        for  key in request :
            signed.append(key)
            signing_string += str(request[key])
        signing_string += nonce + self.secret
        request["signature"] ={
                    "signed":",".join(signed),
                    "token": sha1(signing_string ).hexdigest(),
                    "nonce" : nonce}
        return request
